### ISSUE TITLE
networking broken, /etc/xen/scripts/vif-route-qubes fails in Whonix when IPv6 is enabled

### ISSUE THREAD

**Issue by adrelanos** [QubesOS/qubes-issues#5110]
This issue affects Qubes-Whonix and was triaged by @marmarek, see:
http://forums.whonix.org/t/whonix-gateway-not-reachable/7484/21

@marmarek 
> Configuring `vif*` interfaces is responsibility of `/etc/xen/scripts/vif-route-qubes` script called by `xendriverdomain` service, not udev. It fails, with logs in `/var/log/xen/xen-hotplug.log` . I see a single error there:
> 
> ```
> RTNETLINK answers: Permission denied
> ```
> 
> After adding `set -x` at the beginning of the script, I see it’s on a try to set IPv6 address. Probably because IPv6 is disabled in Whonix.
> Indeed I have enabled IPv6 in my system globally (https://www.qubes-os.org/doc/networking/#ipv6). The same page explains how to disable it for particular VM (and others connected to it). So, I’ve executed `qvm-features sys-whonix ipv6 ''` , restarted sys-whonix and now it works.
> 
> 

Sounds related:
https://github.com/QubesOS/qubes-issues/issues/3961

**Comment by awokd** [comment]
Should the title say "enabled" vs. "disabled"?

**Comment by marmarek** [comment]
@adrelanos what is the status of IPv6 in Tor and Whonix?

If it isn't there at all, maybe we should force disabling IPv6 for Whonix Gateway (and child VMs)? For example using https://github.com/QubesOS/qubes-core-admin-addon-whonix/

**Comment by adrelanos** [comment]
> If it isn't there at all, maybe we should force disabling IPv6 for Whonix Gateway (and child VMs)? For example using https://github.com/QubesOS/qubes-core-admin-addon-whonix/

I am ok with that. But better would be... Is it possible to fix `/etc/xen/scripts/vif-route-qubes` instead? That would be more future proof while keeping the door more open for experimenting / contributions.

> what is the status of IPv6 in Tor and Whonix?

Tor [TransparentProxy](https://trac.torproject.org/projects/tor/wiki/doc/TransparentProxy) doesn't exist for IPv6 and/or nftables.

This is still the latest status:
https://phabricator.whonix.org/T509#9147

**Comment by marmarek** [comment]
> Is it possible to fix `/etc/xen/scripts/vif-route-qubes` instead?

Yes, it would be possible to avoid vif-route-qubes failing in this case. But if IPv6 is enabled, connected VMs would expect it to work, or at least return ICMP errors if it doesn't. Otherwise every single connection to a IPv6-capable host would suffer initial timeouts (sometimes 30-60s). This especially applies to non-Whonix VMs, as I guess Whonix Ws also has IPv6 disabled.
If IPv6 is disabled in the kernel, VMs won't get ICMP error about its unavailability and many connections would take much longer.
In short: just fixing `vif-route-qubes` will trade no connection at all for long waiting times for many connections. At least for non-Whonix VMs.
Some alternative would be to enable IPv6 support in kernel, but filter it at iptables/nftables level (using REJECT, not DROP).

**Comment by adrelanos** [comment]
Marek Marczykowski-Górecki:
>> Is it possible to fix `/etc/xen/scripts/vif-route-qubes` instead?
> 
> Yes, it would be possible to avoid vif-route-qubes failing in this case. But if IPv6 is enabled, connected VMs would expect it to work, or at least return ICMP errors if it doesn't. Otherwise every single connection to a IPv6-capable host would suffer initial timeouts (sometimes 30-60s). This especially applies to non-Whonix VMs,

> In short: just fixing `vif-route-qubes` will trade no connection at
all for long waiting times for many connections. At least for non-Whonix
VMs.


Maybe some settings / status file could influence if it should fail open
or closed?

> as I guess Whonix Ws also has IPv6 disabled.


No, it doesn't. IPv6 is enabled in ws. And it's functional. Tor can
tunnel IPv6 through IPv4. https://ipv6-test.com/ works in Tor Browser.

> Some alternative would be to enable IPv6 support in kernel, but filter it at iptables/nftables level (using REJECT, not DROP).


Btw whonix-firewall does REJECT too.

gw /etc/sysctl.d/ipv6-disable.conf does

net.ipv6.conf.all.disable_ipv6 = 1

**Comment by marmarek** [comment]
> No, it doesn't. IPv6 is enabled in ws. And it's functional. Tor can tunnel IPv6 through IPv4. https://ipv6-test.com/ works in Tor Browser.

But that's over SOCKS, no? How would that work otherwise if you disable IPv6 in Whonix-Gw?

>  In short: just fixing `vif-route-qubes` will trade no connection at all for long waiting times for many connections. At least for non-Whonix VMs.

Hmm, maybe the failure mode could be:
 - log an error message
 - add REJECT firewall rule to avoid timeouts (to support also the cases where it isn't there independently)

**Comment by adrelanos** [comment]
Marek Marczykowski-Górecki:
>> No, it doesn't. IPv6 is enabled in ws. And it's functional. Tor can tunnel IPv6 through IPv4. https://ipv6-test.com/ works in Tor Browser.
> 
> But that's over SOCKS, no?


Yes.

> Hmm, maybe the failure mode could be:
>  - log an error message
>  - add REJECT firewall rule to avoid timeouts (to support also the cases where it isn't there independently)


Yes.

**Comment by marmarek** [comment]
>  - add REJECT firewall rule to avoid timeouts (to support also the cases where it isn't there independently)

In fact, if IPv6 is disabled, no IPv6 firewall rule is evaluated, so it won't help.
